composition-js/src/hints.ts

```diff
@@ -1,164 +1,186 @@
 import { SubgraphASTNode } from "@apollo/federation-internals";
 import { printLocation } from "graphql";
 
-export class HintID {
-  constructor(
-    readonly code: string,
-    readonly description: string,
-    readonly mainElementDescription: string
-  ) {
-  }
+export enum HintLevel {
+  WARN = 60,
+  INFO = 40,
+  DEBUG = 20,
+}
 
-  toString(): string {
-    return `${this.code}: ${this.description}`;
-  }
+export type HintCodeDefinition = {
```

Comment:
I think that you'll find that we want to be able to get the english readable level from the HintCodeDefinition. I'd probably go with something like this, though it would involve changing the HintCodeDefinition back to a class and using a constructor

```typescript
const levels = {
  fatal: 60,
  error: 50,
  warn: 40,
  info: 30,
  debug: 20,
  trace: 10
};

type HintLevel = keyof typeof levels;

export class HintCodeDefinition {
  levelNum: (level: HintLevel) => number = (level) => levels[level];
  constructor(readonly code: string, readonly level: HintLevel, readonly description: string) {}
}
 
const INCONSISTENT_BUT_COMPATIBLE_FIELD_TYPE = new HintCodeDefinition(
  'INCONSISTENT_BUT_COMPATIBLE_FIELD_TYPE',
  'info',
  'Indicates that a field does not have the exact same types in all subgraphs, but that the types are "compatible"'
    + ' (2 types are compatible if one is a non-nullable version of the other, a list version, a subtype, or a'
    + ' combination of the former).',
);
```

Reply:
>  english readable level from the HintCodeDefinition

Isn't it what `HintLevel[def.code]` does? I'm probably brain-farting this but I don't see which advantage the solution above brings (not that I'm against classes :D)?

Reply:
Sorry, I'm not seeing how you go from the value to the name (and specifically how rover will do it). If there is a way that I'm not seeing, feel free to ignore.

Reply:
Oh, I see what you mean. Forgot we were "blindly" serializing JS object to json for rover.

I'll deal with that. But thinking of maybe have level be a `{ value: HintLevel, name: string }`. Kind of made sense to me to talk of the level as being fundamentally an integer value, but some human-readable name attached for convenience.

Reply:
Pushed a change for this. Kept the enum cause it kind of feel appropriate for this to me.

Side-note: I noticed your snippet above included more levels than I have in the patch. I do prefer avoid to add too many levels until it genuinely adds value, but there is of course plenty of space to add more later if the need does arise.

Reply:
On the extra levels thing, I copied some code from `pino`, wasn't suggesting that we actually add all those levels now.